### ISSUE TITLE
How to timeout/cancel query on MySQL backend

### ISSUE THREAD

**Issue by redboltz** [SOCI/soci#690]
I'm using soci 3.2 with mysql backend.

The server side consists two nodes that runs mysqld.
Let's say node A (master) and node B (slave).

The client is connecting to node A. When fail-over happens node B promotes to master.
At that time, node A may not send tcp connection reset or EOF. 
So the client may not notice disconnection.

That behavior is observed AWS RDS mysql. Our team asked Amazon and got response. The response said that the client should disconnect and connect again.

Here is sending query example from http://soci.sourceforge.net/doc/3.2/

```cpp
            // fail-over happens here
            sql << "select phone from phonebook where name = :name",
                into(phone, ind), use(name);
```

If the fail-over happens the comment line, the query execution is blocked.
And I couldn't find the way to cancel execution or set query execution timeout.

NOTE: If the client detects TCP disconnection, soci throws exception. In this case, there is no problem.

Is there any way to do that? 
Or is there any better approach for that situation?

**Comment by redboltz** [comment]
My first solution was setting the timer and when the timer is fired, call `sql.close()`. 
But it seems to be a bad idea because of thread-safety.

http://soci.sourceforge.net/doc/3.2/multithreading.html said that

> The general rule for multithreading is that SOCI classes are not thread-safe, meaning that their instances should not be used concurrently by multiple threads.

During the following code is blocking, there is no way to call `sql.close()` from the same thread.

```cpp
            sql << "select phone from phonebook where name = :name",
                into(phone, ind), use(name);
```

**Comment by redboltz** [comment]
I noticed that #3 could solve the issue.
But according to https://github.com/SOCI/soci/pull/3#issuecomment-330001567, #3 is pretty old. In addition, it contains three different features that setting charset, reconnect, and timeout. And setting charset has already been merged as the different pull request.

I implement timeout support based on up-to-date code. I sent it as #691.

**Comment by redboltz** [comment]
Amazon support said that, RDS mysql fail-over happens not only actual failure situation but also normal operation. I guess that they do virtual machine resource acquisition control. And as I mentioned, when fail-over happens, mysql client like SOCI doesn't handle it. It's critical problem for me.
If #691 would be merged and would include 4.0 milestone, I would appreciate.

**Comment by mloskot** [comment]
@redboltz Thanks for the PR. See my comments there. If it is merged, it will be part of 4.0.
